src/main/kotlin/com/adyen/checkout/api/WebhookResource.kt

```diff
@@ -38,35 +38,35 @@ class WebhookResource @Autowired constructor(@Value("\${ADYEN_HMAC_KEY}") key: S
      */
     @PostMapping("/webhooks/notifications")
     fun webhooks(@RequestBody notificationRequest: NotificationRequest): ResponseEntity<String> {
-        notificationRequest.notificationItems.forEach(
-            Consumer { item: NotificationRequestItem ->
-                // We recommend validate HMAC signature in the webhooks for security reasons
-                try {
-                    if (HMACValidator().validateHMAC(item, hmacKey)) {
-                        log.info(
-                            """
-                                Received webhook with event {} :
-                                Merchant Reference: {}
-                                Alias : {}
-                                PSP reference : {}
-                                """.trimIndent(),
-                            item.eventCode,
-                            item.merchantReference,
-                            item.additionalData["alias"],
-                            item.pspReference
-                        )
-                    } else {
-//                         invalid HMAC signature: do not send [accepted] response
-                        log.warn("Could not validate HMAC signature for incoming webhook message: {}", item)
-                        throw RuntimeException("Invalid HMAC signature")
-                    }
-                } catch (e: SignatureException) {
-                    log.error("Error while validating HMAC Key", e)
+	// JSON and HTTP POST notifications always contain a single NotificationRequestItem object, see https://docs.adyen.com/development-resources/webhooks/understand-notifications#notification-structure
+        notificationRequest.notificationItems.firstOrNull()?.let { item: NotificationRequestItem ->
+            // We always recommend validating HMAC signature in the webhooks for security reasons, see https://docs.adyen.com/development-resources/webhooks/verify-hmac-signatures
+            try {
+                if (HMACValidator().validateHMAC(item, hmacKey)) {
+                    log.info(
+                        """
+                            Received webhook with event {} :
+                            Merchant Reference: {}
+                            Alias : {}
+                            PSP reference : {}
+                            """.trimIndent(),
+                        item.eventCode,
+                        item.merchantReference,
+                        item.additionalData["alias"],
+                        item.pspReference
+                    )
+                } else {
+                    // Invalid HMAC signature: do not send [accepted] response
+                    log.warn("Could not validate HMAC signature for incoming webhook message: {}", item)
+                    throw RuntimeException("Invalid HMAC signature")
                 }
+            } catch (e: SignatureException) {
+                log.error("Error while validating HMAC Key", e)
```

Comment:
Here we should throw `RuntimeException` otherwise the exception is only logged but the workflow continues and sends the `[accepted]' response. I think we have the same problem in the Java app

Reply:
Good catch. Will rework this